Telink b95 new patches

haiwentelink opened the pull request:
fix the old patches

andriy-bilynskyy commented:
The following west manifest projects have been modified in this Pull Request:

| Name | Old Revision | New Revision | Diff |
| ---- | ------------ | ------------ |------|
| hal_telink | https://github.com/telink-semi/hal_telink/commit/1c110b892f438ed491d480ea8fd0f4eb8f5ecfe1 (telink-b95-new) | https://github.com/telink-semi/hal_telink/commit/3cae467b96be07c142fa807bd3c7f900428d3674 (develop) | [telink-semi/hal_telink@1c110b89..3cae467b](https://github.com/telink-semi/hal_telink/compare/1c110b892f438ed491d480ea8fd0f4eb8f5ecfe1..3cae467b96be07c142fa807bd3c7f900428d3674) |

*Note: This message is automatically posted and updated by the Manifest GitHub Action.*

s07641069 commented on boards/riscv/tlsr9258a/tlsr9258a_defconfig:
```diff
@@ -11,3 +11,4 @@ CONFIG_UART_INTERRUPT_DRIVEN=y
 CONFIG_SERIAL=y
 CONFIG_CONSOLE=y
 CONFIG_UART_CONSOLE=y
+CONFIG_TELINK_B9X_2_WIRE_SPI_ENABLE=y
```
with 2 wire you will fail some of certification tests
As well OTA will be not stable

haiwentelink replied:
why we change the spi to 2 wire ,because after poweroff , it will have low rate to revert back 2 wire , so we need to start with 2-wire setting , and then detect the status in  soc_b9x_check_flash to set back to 4 wire . 
so we have two way to slove this problem . 
1. set the config of the     CONFIG_TELINK_B9X_2_WIRE_SPI_ENABLE=y
2. change the cstart settings , just set the 4-wire option to 2 wire value . 
just like that :
                                                              #if defined(CONFIG_TELINK_B9X_2_WIRE_SPI_ENABLE) && CONFIG_TELINK_B9X_2_WIRE_SPI_ENABLE
	.word (0x030020A8)              #/ 2 - wire SPI interface
#else
	.word (0x030020A8)              #/ 4 - wire SPI interface will set in the soc_b9x_check_flash
#endif                     
 
        And I aggree your opinion ,  don't set this define .
        But if we want the define  CONFIG_TELINK_B9X_2_WIRE_SPI_ENABLE  take effect ,should add the define to disable the 4-wire func init in the func of 
                                                                       /* API implementation: driver initialization */
static int flash_b9x_init(const struct device *dev)
{
	struct flash_b9x_data *dev_data = dev->data;

	k_mutex_init(&dev_data->flash_lock);

	flash_change_rw_func(flash_4read, flash_quad_page_program); // this func should add special proc to exclude if the 2-wire take effect .

	return 0;
}

haiwentelink replied:
something like that to control to force use 2-wire 
#if !CONFIG_TELINK_B9X_2_WIRE_SPI_ENABLE
flash_change_rw_func(flash_4read, flash_quad_page_program);
#endif

s07641069 replied:
ok now its clear for me.
Anyway option to use "CONFIG_TELINK_B9X_2_WIRE_SPI_ENABLE=y" is not correct.
When we set 2 wire - we have to use 2 wire only.
I think option 2 is mach clear for understanding and just leave comment for 4 wire like you shown
.word (0x030020A8) #/ 4 - ....
And in place where you set real 4 wire make it for B95 only in flash driver.

s07641069 commented on drivers/adc/adc_b9x.c:
```diff
@@ -21,6 +21,20 @@ LOG_MODULE_REGISTER(adc_b9x, CONFIG_ADC_LOG_LEVEL);
 #include <adc.h>
 #include <zephyr/drivers/pinctrl.h>
 
+#if CONFIG_SOC_RISCV_TELINK_B95
+#define adc_input_pin_def_e     adc_input_pin_e
+#endif
+
+#if CONFIG_SOC_RISCV_TELINK_B95
```
could we use one #if for both changes?

haiwentelink replied:
OK, will merge into one .

s07641069 commented on drivers/adc/adc_b9x.c:
```diff
@@ -91,9 +105,11 @@ static adc_input_pin_def_e adc_b9x_get_pin(uint8_t dt_pin)
 	adc_input_pin_def_e adc_pin;
 
 	switch (dt_pin) {
+#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B95
```
why not needed for B92?

haiwentelink replied:
For B92 don't have PB0 (drivers/B92/adc.h : line 90 ):
typedef enum{
	ADC_GPIO_PB1 = GPIO_PB1 | (0x2<<12),
	ADC_GPIO_PB2 = GPIO_PB2 | (0x3<<12),
	ADC_GPIO_PB3 = GPIO_PB3 | (0x4<<12),
	ADC_GPIO_PB4 = GPIO_PB4 | (0x5<<12),
	ADC_GPIO_PB5 = GPIO_PB5 | (0x6<<12),
	ADC_GPIO_PB6 = GPIO_PB6 | (0x7<<12),
	ADC_GPIO_PB7 = GPIO_PB7 | (0x8<<12),
	ADC_GPIO_PD0 = GPIO_PD0 | (0x9<<12),
	ADC_GPIO_PD1 = GPIO_PD1 | (0xa<<12),
}adc_input_pin_def_e;

s07641069 commented on drivers/entropy/entropy_b9x_trng.c:
```diff
@@ -6,7 +6,11 @@
+#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
 #include <trng.h>
+#elif CONFIG_SOC_RISCV_TELINK_B95
+#include <lib/include/trng/trng.h>
```
Add "lib/include/trng" to CMake include list and revert this change

haiwentelink replied:
will add the lib/include /trng into cmake list .

s07641069 commented on drivers/ieee802154/ieee802154_b9x.c:
```diff
@@ -6,7 +6,11 @@
+#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
 #include "rf.h"
+#elif CONFIG_SOC_RISCV_TELINK_B95
+#include "lib/include/rf/rf_common.h"
```
same Add "lib/include/rf" to CMake include list and revert this change

haiwentelink replied:
will add the lib/include /rf into cmake list .

s07641069 commented on drivers/ieee802154/ieee802154_b9x.c:
```diff
@@ -1051,9 +1055,11 @@ static int b9x_stop(const struct device *dev)
-#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
 		rf_reset_dma();
+#elif CONFIG_SOC_RISCV_TELINK_B95
+		rf_radio_reset();
```
just as improvement we can create "rf_radio_reset" function for b91 and b92 where put rf_baseband_reset(); rf_reset_dma(); to avoid additional #if

haiwentelink replied:
I think it's for pm setting , we have not proc yet , it's need to reserve for B95 pm proc later .

s07641069 replied:
ok

s07641069 commented on drivers/ieee802154/ieee802154_b9x_frame.c:
```diff
@@ -607,8 +608,8 @@ b9x_ieee802154_get_data(const uint8_t *payload,
 #elif CONFIG_SOC_RISCV_TELINK_B95
-#include <ske.h>
-#include <ske_portable.h>
+#include <lib/include/ske/ske.h>
+#include <lib/include/ske/ske_portable.h>
```
same "lib/include/ske/" can be removed after move into CMake

haiwentelink replied:
will add the lib/include /ske into cmake list .

s07641069 commented on soc/riscv/riscv-privilege/telink_b9x/soc.c:
```diff
@@ -315,14 +327,51 @@ void soc_b9x_restore(void)
+#if CONFIG_SOC_RISCV_TELINK_B95
+
+#include "flash/flash_common.h"
+#include "lib/include/flash_base.h"
```
"lib/include" can be removed

haiwentelink replied:
will add the lib/include into cmake list .

s07641069 commented on soc/riscv/riscv-privilege/telink_b9x/soc.c:
```diff
@@ -4,11 +4,20 @@
+#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
 #include "rf.h"
+#elif CONFIG_SOC_RISCV_TELINK_B95
+#include "lib/include/rf/rf_common.h"
```
lib/include can be removed and optimized to use only one incluse for all SoCs

haiwentelink replied:
will add the lib/include /rf into cmake list .

s07641069 commented on soc/riscv/riscv-privilege/telink_b9x/soc.c:
```diff
@@ -331,9 +380,17 @@ static int soc_b9x_check_flash(void)
+	/* Enable 4x SPI read and write */
+
+	if (flash_set_4line_read_write(SLAVE0, mid) != 1) {
+		/*printk("set 4line fail");*/
```
remove commented code

haiwentelink replied:
OK, will delete the comment code.

s07641069 commented on soc/riscv/riscv-privilege/telink_b9x/soc.c:
```diff
@@ -4,11 +4,20 @@
 #include <sys.h>
+#elif CONFIG_SOC_RISCV_TELINK_B95
+#include <lib/include/sys.h>
+#include <zephyr/sys/printk.h>
```
remove unused #include <zephyr/sys/printk.h>

haiwentelink replied:
In B92 , it will not have warning , but in B95 , if not add the <zephyr/sys/printk>, it will have warning , but i don't know why , need time to make it clear .
"/home/matter/zephyrproject/zephyr/soc/riscv/riscv-privilege/telink_b9x/soc.c:414:17: warning: implicit declaration of function 'printk' [-Wimplicit-function-declaration]
  414 |                 printk("!!! flash error: expected (.dts) %u, actually %u\n",
      |                 ^~~~~~
[577/676] No configure step for 'chip-gn'"

s07641069 replied:
ok resolve this warning and after revert this change 
as well we can use one sys.h incluse for all SoCs here

s07641069 commented on soc/riscv/riscv-privilege/telink_b9x/soc.c:
```diff
@@ -4,11 +4,20 @@
+#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
 #include <sys.h>
+#elif CONFIG_SOC_RISCV_TELINK_B95
+#include <lib/include/sys.h>
```
"lib/include" can be removed and we can use one incluse for all SoCs

haiwentelink replied:
will add the lib/include  into cmake list .

s07641069 commented on samples/drivers/adc/boards/tlsr9258a.overlay:
```diff
@@ -0,0 +1,26 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2022 Telink Semiconductor (Shanghai) Co., Ltd.
```
use correct year

haiwentelink replied:
OK

s07641069 commented on dts/bindings/adc/adc-controller.yaml:
This is common (not Telink) file.
revert this file and create Telink own yaml file for same purpose.
We already did this for other drivers.

haiwentelink replied:
will revert , and use another exist setting "zephyr,gain = "ADC_GAIN_1";", and add the proc in adc_b9x.

                adc_code = adc_b9x_get_code();
		if (!data->differential) {
			/* Sign bit is not used in case of single-ended configuration */
			adc_code = adc_code * 8;

			/* Do not return negative value for single-ended configuration */
			if (adc_code < 0) {
				adc_code = 0;
			}
		}

s07641069 replied:
you can use this way or create Telink yaml file instead of modifying common yaml file

s07641069 commented on drivers/adc/adc_common.c:
This is common (not Telink) file.
revert this file and rework to apply changes into Telink only files.

haiwentelink replied:
will revert

s07641069 commented on include/zephyr/drivers/adc.h:
This is common (not Telink) file.
revert this file and rework to apply changes into Telink only files.

haiwentelink replied:
will revert

andriy-bilynskyy commented:
PR seems fine. But plese squash your commits. There is no any reasons to keep middle sources version where builds are not tested.

s07641069 commented:
@haiwentelink
next process:
1. merge https://github.com/telink-semi/zephyr_hal_telink_b91_ble_lib/pull/15 (approved) 
2. update in https://github.com/telink-semi/hal_telink/pull/86 hash of lib (do again force push to have 1 commit) approved
3. update in this PR (again force push) pull into hash of HAL (for west)
4. after I will approve this PR and merge it

haiwentelink committed:
telink: B95: clean code and adjust adc demo.

- clean code to make code more clear.
- change the 4-wire setting.
- change adc setting part , use init setting and adjust the code.
- fix the warning of "security not found".
- fix the warning of "security not found".

Signed-off-by: Haiwen Xia <haiwen.xia@telink-semi.com>
